File: Math/PermutationSequence.py
# LeetCode 60 - Permutation Sequence
# Time Complexity: O(n^2) | Space Complexity: O(n)


class Solution:
    def getPermutation(self, n: int, k: int) -> str:
        answer: list[str] = []
        visited = [False] * (n + 1)

        for index in range(n):
            factorial = 1
            for value in range(1, n - index):
                factorial *= value

            for value in range(1, n + 1):
                if not visited[value]:
                    if k > factorial:
                        k -= factorial
                    else:
                        answer.append(str(value))
                        visited[value] = True
                        break

        return "".join(answer)

File: Math/PermutationSequence.java
// LeetCode 60 - Permutation Sequence
// Time Complexity: O(n^2) | Space Complexity: O(n)
public class PermutationSequence {
    public String getPermutation(int n, int k) {
        StringBuilder answer = new StringBuilder();
        boolean[] visited = new boolean[n + 1];

        for (int index = 0; index < n; index++) {
            int factorial = 1;
            for (int value = 1; value < n - index; value++) {
                factorial *= value;
            }

            for (int value = 1; value <= n; value++) {
                if (!visited[value]) {
                    if (k > factorial) {
                        k -= factorial;
                    } else {
                        answer.append(value);
                        visited[value] = true;
                        break;
                    }
                }
            }
        }

        return answer.toString();
    }
}
